### PR TITLE
Allow to transform Trajectories

### LINKED ISSUE #565: Is there a way to repeat a structure and then animate the structure?
Hello,

I am wondering if it would be possible to repeat a structure first, and then use animate_structure() ? Currently, it looks like it is not possible. Any thoughts?

### PR DESCRIPTION
**Pull request by pmrv**
Use it in animate_structure to allow for repeated structures.

I've just quickly checked this manually.  @skatnagallu, please give it another check as well.  Also if you have time/interest I think it'd be a good exercise to add simple unit tests to `TransformTrajectory`. ;) 

We might think about exposing the transform function from `animate_structure` to allow for more fancy visualizations, but I'll leave until someone has an application.

Fixes #565.

### CODE REVIEW

**Review by sudarsan-surendralal — COMMENTED** [review]
Looks super elegant! The only thing we need is tests for `TransformTrajectory` just to be sure

**Comment by skatnagallu** [comment]
Ok. I will give it a try. It will be a good exercise for me. :) Thanks for solving this so quickly.

**Comment by skatnagallu** [comment]
Hello, Sorry that I got back to this a bit late. I am trying to write my tests and run them on my local pc. But I keep getting this error. 
ImportError: cannot import name 'state' from 'pyiron_base'

**Comment by pmrv** [comment]
> Hello, Sorry that I got back to this a bit late. I am trying to write my tests and run them on my local pc. But I keep getting this error. ImportError: cannot import name 'state' from 'pyiron_base'

Sounds like your local environment is out of date.  If you're testing inside a conda environment run `conda env update -f .ci_support/environment.yml` after activating the environment.

**Comment by skatnagallu** [comment]
I am working in pycharm. I am not sure how I can update the environment here.

**Comment by pmrv** [comment]
I'm not familiar with pycharm, but I suppose you want [this](https://www.jetbrains.com/help/pycharm/conda-support-creating-conda-virtual-environment.html#conda-requirements).

**Comment by skatnagallu** [comment]
I added the test, but I am not sure if it is doing what the transform trajectory is supposed to do. Could you have a look @pmrv ?

**Comment by sudarsan-surendralal** [comment]
@skatnagallu , I think you forgot to push the changes here

**Comment by pmrv** [comment]
It seems you forgot to push your commits.  Do `git push` or `git push -o origin trajrep` from your repository root.

Generally the tests should just do some basic sanity checks, like
- [ ] transformed trajectories should be of the same length
- [ ] check that transformations are properly applied, e.g. check once that transforming with the identity function `lambda x: x` returns the same structures and check once something like `lambda x: x.repeat(2)` and then check that the length of each structure is multiplied by 8.

**Comment by skatnagallu** [comment]
> It seems you forgot to push your commits. Do `git push` or `git push -o origin trajrep` from your repository root.
> 
> Generally the tests should just do some basic sanity checks, like
> 
> * [ ]  transformed trajectories should be of the same length
> * [ ]  check that transformations are properly applied, e.g. check once that transforming with the identity function `lambda x: x` returns the same structures and check once something like `lambda x: x.repeat(2)` and then check that the length of each structure is multiplied by 8.

Yes, so this is what I checked, but the lengths do not match. I am not sure , if I have applied the transformations properly.

**Comment by skatnagallu** [comment]
By the way, I accidentally created a new branch called traj_rep and pushed my changes there. How can I delete this?

File: tests/atomistics/job/test_transform_trajectory.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+        shutil.copy(os.path.join(tests_loc, "static/lammps_test_files/get_structure_test.tar.gz"), cwd)
+        shutil.copy(os.path.join(tests_loc, "static/lammps_test_files/export.csv"), cwd)
+        self.project.unpack("get_structure_test")
+        job = self.project.load("inter_calculator")
+        traj = job.trajectory()
+        self.assertIsInstance(traj, Trajectory)
+        self.assertEqual(len(traj), len(self.job.output.positions))
+        traj = traj.transform(lambda s: s.repeat(2))
+        self.assertIsInstance(traj, TransformTrajectory)
+        self.assertEqual(len(traj), 8*len(self.job.output.positions))
```
Comment (pmrv) [inline tests/atomistics/job/test_transform_trajectory.py]: If the structures are repeated then the length of each structure goes by 8 not the whole trajectory.  So you probably need something like

```suggestion
for s1, s2 in zip(traj.iter_structures(), self.job.iter_structures()):
    self.assertEqual(len(s1) * 8, len(s2), "Structure not properly transformed!")
```

and something similar for the identify function.

**Comment by pmrv** [comment]
> By the way, I accidentally created a new branch called traj_rep and pushed my changes there. How can I delete this?

Go the main repository page, click on `branches` on top then find `traj_rep` under `Your Branches` then delete on the trash symbol.

**Comment by skatnagallu** [comment]
@pmrv All tests run perfectly. I do not see any more problems. The branch can be merged.

File: tests/atomistics/job/test_transform_trajectory.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+        traj = self.job.trajectory()
+        traj = traj.transform(lambda s: s.repeat(2))
+        self.assertEqual(len(traj.get_structure()), 8*len(self.job.get_structure()))
+        traj = self.job.trajectory()
+        traj = traj.transform(lambda s: s.repeat(1))
+        self.assertEqual(len(traj.get_structure()), 1*len(self.job.get_structure()))
+        traj = self.job.trajectory()
+        traj = traj.transform(lambda s: s.repeat(2))
+        for i in traj.iter_structures():
+            self.assertEqual(len(i), 8*len(self.job.get_structure()))
```
Comment (pmrv) [inline tests/atomistics/job/test_transform_trajectory.py]:
```suggestion
        for s1, s2 in zip(traj.iter_structures(), self.job.iter_structure()):
            self.assertEqual(len(s1), 8*len(s2))
```

Comment (skatnagallu) [reply tests/atomistics/job/test_transform_trajectory.py]: I think it should be `for s1, s2 in zip(traj.iter_structures(), self.job.iter_structures()):`

Comment (pmrv) [reply tests/atomistics/job/test_transform_trajectory.py]: Oh yeah, good catch. :|

File: tests/atomistics/job/test_transform_trajectory.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+class ToyAtomisticJob(AtomisticGenericJob, ABC):
+
+    def _check_if_input_should_be_written(self):
+        return False
+
+    def run_static(self):
+        self.save()
+        self.status.running = True
+        self.status.finished = True
+        self.to_hdf()
```
Comment (niklassiemer) [inline tests/atomistics/job/test_transform_trajectory.py]: This should not be required. The save should already be called by default! 
```suggestion
    def run_static(self):
        self.status.finished = True
        self.to_hdf()
```
should be enough.

**Review by niklassiemer — COMMENTED** [review]
I just had a quick glance at it :)

File: tests/atomistics/job/test_transform_trajectory.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
+import os
+import shutil
+import unittest
+from abc import ABC
+import numpy as np
+from pathlib import Path
+from pyiron_base._tests import TestWithCleanProject
+from pyiron_base import ProjectHDFio
+from pyiron_atomistics.atomistics.job.atomistic import AtomisticGenericJob, Trajectory, TransformTrajectory
+from pyiron_atomistics.atomistics.structure.atoms import Atoms, CrystalStructure
```
Comment (sudarsan-surendralal) [inline tests/atomistics/job/test_transform_trajectory.py]: Avoid importing modules you don't actually use 

```suggestion
from pyiron_atomistics.atomistics.structure.atoms import CrystalStructure
```

File: pyiron_atomistics/atomistics/job/atomistic.py [hunk 5]
```diff
@@ -5,7 +5,9 @@
 import copy
 import warnings
 import numpy as np
+import numbers
 import os
+from typing import Callable, Union, Tuple
```
Comment (sudarsan-surendralal) [inline pyiron_atomistics/atomistics/job/atomistic.py:10]: You could avoid the re-import of `Union` in L27

File: tests/atomistics/job/test_transform_trajectory.py [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+
+class ToyAtomisticJob(AtomisticGenericJob, ABC):
```
Comment (sudarsan-surendralal) [inline tests/atomistics/job/test_transform_trajectory.py]: To avoid code duplication, you could import this class from `test_atomistic.py`. (You could implement @niklassiemer's suggestion there)

**Review by sudarsan-surendralal — COMMENTED** [review]
Just a few suggestions from my side based on what codeacy highlighted (https://app.codacy.com/gh/pyiron/pyiron_atomistics/pullRequest?prid=9080376)

**Comment by skatnagallu** [comment]
I have implemented all of your suggesstions.

File: tests/atomistics/job/test_transform_trajectory.py [hunk 7]
```diff
@@ -0,0 +1,47 @@
+import os
+import shutil
+import unittest
+from pathlib import Path
+from pyiron_base._tests import TestWithCleanProject
+from pyiron_base import ProjectHDFio
+from pyiron_atomistics.atomistics.job.atomistic import Trajectory
+from pyiron_atomistics.atomistics.structure.atoms import CrystalStructure
+from test_atomistic import ToyAtomisticJob
```
Comment (sudarsan-surendralal) [inline tests/atomistics/job/test_transform_trajectory.py]: To fix the tests:

```suggestion
from .test_atomistic import ToyAtomisticJob
```

**Review by sudarsan-surendralal — COMMENTED** [review]
Just a minor change required. Otherwise it looks good from my side (although the tests unrelated to what you developed seem to fail)

**Comment by skatnagallu** [comment]
Can we merge and close this then?

**Comment by sudarsan-surendralal** [comment]
@skatnagallu I'm a bit unsure about why the Sphinx tests suddenly fail. @samwaseda any idea why this happens? (see below). If we are sure they are unrelated I gues we can merge this

```

======================================================================
FAIL: test_get_scf_group (sphinx.test_base.TestSphinx)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/pyiron_atomistics/pyiron_atomistics/tests/sphinx/test_base.py", line 336, in test_get_scf_group
    self.assertEqual(test_scf, ref_scf)
AssertionError: Group({'rhoMixing': '1.0', 'spinMixing': [159 chars][])}) != {'rhoMixing': '1.0', 'spinMixing': '1.0',[138 chars]: {}}

======================================================================
FAIL: test_write_input (sphinx.test_base.TestSphinx)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/pyiron_atomistics/pyiron_atomistics/tests/sphinx/test_base.py", line 219, in test_write_input
    self.assertEqual(file_content, lines)
AssertionError: Lists differ: ['//j[842 chars]t = 24.989539079445393;\n', '\tkPoint {\n', '\[1030 chars]}\n'] != ['//j[842 chars]t = 20.0;\n', '\tkPoint {\n', '\t\tcoords = [0[1016 chars]}\n']

First differing element 35:
'\teCut = 24.989539079445393;\n'
'\teCut = 20.0;\n'

```

File: pyiron_atomistics/sphinx/base.py [hunk 8]
```diff
@@ -388,16 +388,16 @@ def load_default_input(self):
         sph.create_group("initialGuess")
         sph.create_group("main")
 
-        self.input.EnCut = 340
+        self.input.EnCut = 20 * RYDBERG_TO_EV
```
Comment (samwaseda) [inline pyiron_atomistics/sphinx/base.py:391]: These changes are not related to the idea of this PR, correct? I'd request them to be done in a different PR. Btw. these changes probably make the tests fail.

Comment (samwaseda) [inline pyiron_atomistics/sphinx/base.py]: These changes are not related to the idea of this PR, correct? I'd request them to be done in a different PR. Btw. these changes probably make the tests fail.

**Review by samwaseda — CHANGES_REQUESTED** [review]
It looks like there are two things happening in this PR - what the title says and the changes of the SPHInX default values. They have to be separated.

File: pyiron_atomistics/sphinx/base.py (hunk 8 above)
Comment (skatnagallu) [reply pyiron_atomistics/sphinx/base.py]: Yes, sorry. I was working on a different branch to do this. But I am not sure how they crept into this branch.

**Review by samwaseda — APPROVED** [review]

**Review by sudarsan-surendralal — APPROVED** [review]
Finally the unit tests pass so feel free to merge this. Thanks @pmrv and @skatnagallu

**Review by skatnagallu — APPROVED** [review]
Ok everything looks good. I will wait for @pmrv to do the honors.
